// laravel-elixir/dist/tasks/TaskReporter.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _gulpUtil = require('gulp-util');

var _gulpUtil2 = _interopRequireDefault(_gulpUtil);

var _cliTable = require('cli-table');

var _cliTable2 = _interopRequireDefault(_cliTable);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var TaskReporter = function () {
    function TaskReporter() {
        _classCallCheck(this, TaskReporter);
    }

    _createClass(TaskReporter, [{
        key: 'report',


        /**
         * Render a single task.
         *
         * @param {Task|null} task
         */
        value: function report(task) {
            var table = this.makeTable();

            this.addRows(table, task ? [task] : Elixir.tasks);

            Elixir.log.message(table.toString());
        }

        /**
         * Create the table rows.
         */

    }, {
        key: 'makeTable',
        value: function makeTable() {
            return new _cliTable2.default({
                head: ['Task', 'Summary', 'Source Files', 'Destination']
            });
        }

        /**
         * Add any number of rows to the table.
         *
         * @param {Table} table
         * @param {array} tasks
         */

    }, {
        key: 'addRows',
        value: function addRows(table, tasks) {
            var _this = this;

            tasks.forEach(function (task) {
                var row = ['mix.' + task.name + '()', task.summary()];

                if (task.src && task.output) {
                    row.push(_this.src(task), task.output.path || task.output);
                }

                table.push(row);
            });
        }

        /**
         * Get a string version of the src files.
         *
         * @param  {Task} task
         * @return {string}
         */

    }, {
        key: 'src',
        value: function src(task) {
            var _this2 = this;

            var src = task.src.path || task.src;

            src = Array.isArray(src) ? src : [src];

            return src.map(function (file) {
                return _this2.colorize(file);
            }).join('\n');
        }

        /**
         * Get a colorized version of the file path,
         * based upon its existence.
         *
         * @param  {string}  file
         * @return {string}
         */

    }, {
        key: 'colorize',
        value: function colorize(file) {
            if (fileExists(file)) {
                return _gulpUtil2.default.colors.green(file);
            }

            return _gulpUtil2.default.colors.bgRed(file);
        }
    }]);

    return TaskReporter;
}();

/**
 * Assert that the given file exists.
 *
 * @param  {string} file
 * @return {boolean}
 */


function fileExists(file) {
    // If this file begins with a !, then the
    // user intends to exclude it from the
    // src set; we're free to ignore it.
    if (file.indexOf('!') == 0) return true;

    return file.match(/\*/) || _fs2.default.existsSync(file);
}

exports.default = TaskReporter;
